### DISCUSSION TITLE
HTML/JS user interfaces

### DISCUSSION THREAD

**Discussion by lucianoiam** [DISTRHO/DPF#290] [Ideas]
Hi @falkTX. I've been working on this for a while https://github.com/lucianoiam/dpf-webui and wondering if it makes sense to integrate it into DPF as one of the UI_TYPEs . While it is overkill for typical plugin work consisting of knobs and switches, It could be interesting for building more app-like complex UIs (thinking of the various JS frameworks out there). I personally dislike webapps but in contrast to let's say Electron, this is not about embracing JS for everything but rather adding some bits of its goodness to the already wild world of plugin UIs. And of course leveraging a full web rendering engine for flexbox, canvas, SVG, etc. It could work for some specific cases while following the DSP/UI separation paradigm DPF enforces.

A more generic, perhaps better integration could involve turning the web view into one of DPF's native Widgets. Then explain the web-based user interface approach in one of the example projects. This breaks dpf-webui into two separate tools: a reusable webview and a JS bridge to DISTRHO::UI.

I think of dpf-webui as a naive but working implementation of a (debatable) good idea which is throwing some higher level language pieces on top of the audio engine core like Bitwig does, or a quite recent project called React-JUCE. While drawing from DPF's spirit of not complicating things.

The interfaces to the system web views were written from scratch to ensure they are safe to run in a plugin context on all platforms.  No dependencies needed.

Being said this, thanks to the DPF team for all the great work.

**Comment by falkTX** [comment]
Hello and thanks for all the work!

I am a bit apprehensive about HTML on plugin UIs, but maybe indeed there is sometimes a good use for them.
The way to go on here that I really recommend, is to try with DPF develop branch and make the WebView stuff be a subclass of TopLevelWidget (this is a new thing in develop branch).
I basically took the things that made the "UI" class special, and put them as a TopLevelWidget. So it can be used not just for plugins but also standalones.

Quick look through the code, I have a few more comments:
1. use typedef's instead of macros, for example for DISTRHO_WEBVIEW_CLASS
2. there is a native API to find scale factor in DPF now. it is only implemented for plugins right now, standalones can use an env var, but best to switch to that
3. switch the webview to inherit from TopLevelWidget instead of UI class
4. it is possible to dlopen the standalone executable, see https://stackoverflow.com/questions/19713517/how-to-self-dlopen-an-executable-binary

**Reply by lucianoiam** [reply]
Thanks for having a look and for the tips! I see some hot progress going on develop and no. 3 sounds like the right way to support the web view.

Off-topic I see a lot of value in the standalone target, I recall reading it was originally created for rapidly testing plugins. Please give it the same treatment as the plugin targets (I think there have been already improvements in that department lately). The only "bottleneck" is JACK which is a bit alien to anything non Linux. Bottleneck because there is so much heavy underlying work for Mac/Win that can get potentially shadowed by the fact, and the double quotes because I love JACK and use it on a daily basis -- but that is not true for my musician friends and probably the average audio-oriented user. Is it a decent idea to bundle a "private, invisible to foreign processes" JACK server into the standalone version? likely the server needs modifications. Or is there any portable audio library that outputs to JACK as an option apart from the system API? the latter sounds interesting as you'd get the best from all worlds. I feel I'm not inventing the wheel but cannot recall seeing that.

**Reply by falkTX** [reply]
I changed DPF recently so that JACK libs/headers are not needed to build the JACK target, it will try to dlopen libjack at the start.
At some point we could get a native dialog popup indicating JACK is not available or started.

The problem with doing something outside of JACK for standalones is that it would need user options to setup device, buffer size, sample rate, midi devices etc..
This goes against the idea of DPF being sorta minimalistic, and I do not plan to add such things at all.
(A separate project will make this possible though, loading DPF plugins and treating the whole deal as standalone https://github.com/falkTX/Chibi)

JACK for audio plugin standalones is fine IMO, I am maintaining it so that it still works on macOS and Windows, so it is not hard to get things working IMO, just needs to be installed and running.

When I speak of standalones I am not just referring to audio plugins though, but using DPF DGL things as actual GUI library for an application (without audio).
Having HTML support means electron-like applications become possible, but without the heavy electron engine behind it.

**Comment by falkTX** [comment]
btw, develop DPF has a "stub" target now too. something that does not setup openGL, cairo or any graphics backend.
Once you have the webview as a toplevelwidget, I will be interested to see if we could make it work with the "stub" backend, so we can skip unnecessary opengl things.

**Comment by lucianoiam** [comment]
DPF minimalism if one of its greatest strengths and agree on that. The plugin shell approach like Chibi sounds a lot better, I was not aware of it. The best of these things and it also applies to JACK is being modular and they should remain the same.

I agree on JACK not hard to get working but er... you know people sometimes are lazy :). I have some ideas to help on that but it would be dishonest to continue the rant because I am lazy myself to implement them. Just bringing up the point that an extra step in the workflow bothers some poeple and that is out of the developers control, especially on those systems where everything is "automatic". No wonder why Pro Tools looks basically the same as 30 years ago. 

In fact dpf-webui was conceived as the starting point for a port of a custom modular system developed in the past based on Python, the hacked Live Live API, HTML/JS, some audio utilities and JACK for inter-app audio. Replicating the setup on someone else's computer proved to be non trivial. It occurred to me that if I tried to (forgive me) glue everything together into a VST plugin it would make it easier to share the thing with bandmates and eventually release it to the general public. So the workflow remains the same for them, start the DAW and forget. Audio routing is done inside the DAW, the visual interface can be casted to the network, etc. It is a very specific case.

I saw the "stub" target yesterday and had a vague idea, the comment confirms I definitely need to dig into it. The web view only needs a pointer to the native parent view, thanks !

**Reply by falkTX** [reply]
well worst case for audio plugin standalones would be a different target, I think SDL would work well.
so we could try JACK if available, and if not fallback to SDL but still keep the jack API intact.

**Reply by lucianoiam** [reply]
Now I think there is no need for a sound configuration panel, if it worked like that then for the user it would be automagical. If they want to get benefited from studio-like facilities the more advanced users will install JACK, while keeping it simple for others (install - run - profit).

**Comment by lucianoiam** [comment]
There is a new branch https://github.com/lucianoiam/dpf-webui/tree/moredpf . The class hierarchy looks like this now:

`TopLevelWidget` ← `BaseWebWidget` ← `PlatformWebWidget`   ( the latter now a typedef, thanks )

I naively tried switching UI_TYPE to stub but the project won't compile so there is clearly something I am overlooking. I am still not familiar with the DPF design; everything built so far was done on top of it and not hooking into it.

The top level UI for this project was implemented as a  derived class of `DISTRHO::UI` ( called `DISTRHO::WebUI` ) because its goal is acting as a user interface to the `Plugin`, ie. call `setParameterValue()`, listen to `parameterChanged()`, and so on. `DISTRHO::WebUI` is a controller for the web widget that basically does two things 1) display the web widget 2) act as a proxy to JS class `DISTRHO_WebUI`, leveraging a minimalistic generic bridge provided by the web widget. The web widget itself is "dumb".

Following your convention, isn't `DISTRHO::WebUI` the top level widget of this design?  why `*WebWidget` should derive from `TopLevelWidget` and not `Widget`? . Is it something related to the graphical backend?

Summarizing in inaccurate terms, this is what I'm thinking of:

1. Have a `DISTRHO::UI` that skips the graphical backend but still provides the raw native window. This could be useful for other blasphemic variations of DPF. Is this what 'stub' is for ?
2. Ideally make the WebWidgets suitable for adding as subwidgets anywhere so they are not limited to what dpf-webui does. There's some potential there even beyond rendering, like scripting or network (the latter needs some adjustments).

```
In file included from ./lib/DPF/distrho/src/DistrhoUI.cpp:17:
./lib/DPF/distrho/src/DistrhoUIPrivateData.hpp:232:29: error: base class has incomplete type
class PluginWindow : public Window
                     ~~~~~~~^~~~~~
./lib/DPF/dgl/Widget.hpp:30:7: note: forward declaration of 'DGL::Window'
class Window;
      ^
In file included from lib/DPF/distrho/DistrhoUIMain.cpp:17:
In file included from ./lib/DPF/distrho/src/DistrhoUI.cpp:17:
./lib/DPF/distrho/src/DistrhoUIPrivateData.hpp:242:76: error: only virtual member functions can be marked 'override'
    void onFocus(const bool focus, const DGL_NAMESPACE::CrossingMode mode) override
                                                                           ^~~~~~~~
...
```

**Reply by lucianoiam** [reply]
FYI just renamed `DISTRHO::WebUI` to `DISTRHO::ProxyWebUI` because it better explains its existence.

**Reply by falkTX** [reply]
> Following your convention, isn't `DISTRHO::WebUI` the top level widget of this design? why `*WebWidget` should derive from `TopLevelWidget` and not `Widget`? . Is it something related to the graphical backend?

I made the TopLevelWidget something that can interact directly with a Window class, as required for plugin UIs.
But intention is to also allow other non-DPF things to become TopLevelWidgets. Cases such as vstgui and imgui, where they already have feature-complete APIs for rendering, sub-widgets etc. TopLevelWidgets can also be fully external windows (via embed)

Having arbitrary widgets (which can be used as children of other widgets) become of type vstgui is very problematic.
It would involve reparenting within reparenting (and the whole thing about keeping position and size consistent) and generally would be too much of a hassle.
This was already seen with NanoVG, where the use of subwidgets becomes very confusing and hard to maintain.

So my idea is to get rid of all complexities by creating a widget class that stands on top of all other widgets.
This one can then be made to work as vstgui, imgui or in your case webgui.
BUT, and this is crucial, it should not be allowed as a "child widget" of any sort.

It is something weird to see, because I do not think any other widget libraries do this.
But it makes sense in this context, where we want to have subwidgets when using more core classes in dpf (imageknob etc, which can use cairo or opengl, position and size internally handled by dpf itself) and we also want to bring in external, widget-libraries-with-widget-logic-already-done stuff.

I think this is the best of both worlds.
People that want to use DPF provided things can do so, but using DPF only for its plugin support and barebones event handling coupled with some other project (vstgui, imgui, etc) is also possible.
IMO DPF should not try to manage stuff like vstgui, it should only provide the small layer needed to translate events and draw things on screen.

I have it commented out for now, but eventually the ExternalUI type can become well supported within DPF, where it embeds and sends events around, very much like you are doing now with webgtk.
Because we do not have to deal with child widgets or the ExternalUI itself possibly being a child widget, we can focus on single class use which is the only supported method anyway.

**Reply by falkTX** [reply]
>     1. Is this what 'stub' is for ?

The "stub" is not known if it is going to work or not, I didnt do any experiments with it yet, only exposed its build.

The theory behind it is that we can create a OS Window via pugl, which has a valid window id that can be used for reparenting.
So that we can couple this together with stuff like vstgui which handles its own drawing internally.
Also so that we can skip initializing opengl or cairo, since the window is going to be reparented anyway.

As an extra step, though not sure if this is possible yet, I might to try get some drawing operations working with pure OS window methods. Think of an X11 window without cairo or opengl surface.
DPF provided classes so far only draw via raw images, so in theory we can have some DPF plugins that draw RGBA things on top of a window directly.
But this last part is more of an experiment than anything else.

**Reply by lucianoiam** [reply]
> I made the TopLevelWidget something that can interact directly with a Window class, as required for plugin UIs.
> But intention is to also allow other non-DPF things to become TopLevelWidgets. Cases such as vstgui and imgui, where they already have feature-complete APIs for rendering, sub-widgets etc. TopLevelWidgets can also be fully external windows (via embed)

Embed like in XEmbed? one of the issues I found while developing dpf-webui is that XEmbed is not supported on Wayland. That does not prevent the web view from working, but ends up in a UX mess because of extra floating windows and gives the web view approach less credibility. A big part of the web view idea is about appearance. 

I yet need to find a way to leverage WebKitGTK without running GTK in the plugin itself and without the need of XEmbed...

> 
> Having arbitrary widgets (which can be used as children of other widgets) become of type vstgui is very problematic.
> It would involve reparenting within reparenting (and the whole thing about keeping position and size consistent) and generally would be too much of a hassle.
> This was already seen with NanoVG, where the use of subwidgets becomes very confusing and hard to maintain.
> 
> So my idea is to get rid of all complexities by creating a widget class that stands on top of all other widgets.
> This one can then be made to work as vstgui, imgui or in your case webgui.
> BUT, and this is crucial, it should not be allowed as a "child widget" of any sort.

Thanks for all the information, it makes sense now. The natural choice for the (now called) `AbstractWebWidget` is deriving from `TopLevelWidget`— good.

The web widget initially used `TopLevelWidget` then switched to `SubWidget` just for the sake of experimentation (true by the time of writing this). So far there have been no issues on any platform but after reading the comments it seems that there are latent side effects depending the context where it is used.

Let's say you have a standard DGL-widgets based plugin, and want  to build a web-based flashy help screen for the plugin, because in that way it is easy to add a contents index, images, etc.  The window would then present the web help, and a native button for closing it, all baked into the same layout but without superposition (think of the toolbar and the actual webview in a browser). In this design the web view usage is completely reduced to rendering, no logic even not for a simple close button.  

Since the TopLevelWidget takes up the whole Window it wouldn't be possible to do it, is that correct?

I am trying to maximize what the web widget can do. Overall I'd rather strip features than allowing incorrect usages of the widget.

> 
> It is something weird to see, because I do not think any other widget libraries do this.
> But it makes sense in this context, where we want to have subwidgets when using more core classes in dpf (imageknob etc, which can use cairo or opengl, position and size internally handled by dpf itself) and we also want to bring in external, widget-libraries-with-widget-logic-already-done stuff.
> 
> I think this is the best of both worlds.

Agree. Making the graphics toolkit optional just follows the framework's spirit. Also there are other frameworks that support the concept.

> People that want to use DPF provided things can do so, but using DPF only for its plugin support and barebones event handling coupled with some other project (vstgui, imgui, etc) is also possible.
> IMO DPF should not try to manage stuff like vstgui, it should only provide the small layer needed to translate events and draw things on screen.
> 
> I have it commented out for now, but eventually the ExternalUI type can become well supported within DPF, where it embeds and sends events around, very much like you are doing now with webgtk.
> Because we do not have to deal with child widgets or the ExternalUI itself possibly being a child widget, we can focus on single class use which is the only supported method anyway.

It is great to see all this evolving!  thanks for all the info.

**Reply by lucianoiam** [reply]
Forgot this.

> The "stub" is not known if it is going to work or not, I didnt do any experiments with it yet, only exposed its build.
> 
> 

In my own quick tries it just fails to build at least on Mac so I couldn't go very far.

> The theory behind it is that we can create a OS Window via pugl, which has a valid window id that can be used for reparenting.

That is everything you'd need for a purely web based plugin. Not so relevant for the web widget as an embeddable widget (I think).

> 
> So that we can couple this together with stuff like vstgui which handles its own drawing internally.
> 
> Also so that we can skip initializing opengl or cairo, since the window is going to be reparented anyway.

Exactly


> 
> 
> As an extra step, though not sure if this is possible yet, I might to try get some drawing operations working with pure OS window methods. Think of an X11 window without cairo or opengl surface.
> 
> DPF provided classes so far only draw via raw images, so in theory we can have some DPF plugins that draw RGBA things on top of a window directly.
> 
> But this last part is more of an experiment than anything else.

Definitely wild but interesting !

**Reply by falkTX** [reply]
> Embed like in XEmbed? one of the issues I found while developing dpf-webui is that XEmbed is not supported on Wayland. That does not prevent the web view from working, but ends up in a UX mess because of extra floating windows and gives the web view approach less credibility. A big part of the web view idea is about appearance.
> 
> I yet need to find a way to leverage WebKitGTK without running GTK in the plugin itself and without the need of XEmbed...

Did you try to force gtk x11 backend?
Since the gtk stuff is in a separate process, this wont conflict with any running tools.
I know you can at least set an environment variable to set gtk backends, maybe possible with code too.

**Reply by falkTX** [reply]
> Since the TopLevelWidget takes up the whole Window it wouldn't be possible to do it, is that correct?

Yes that is correct, top level widgets are meant to always take 100% of the size.
That said.. you can have multiple top level widgets per window. I already use such approach to add a resize handle which is always at the bottom-right of the window.

**Reply by lucianoiam** [reply]
> you can have multiple top level widgets per window. I already use such approach to add a resize handle which is always at the bottom-right of the window.

^^^ this. Now there are no cons in picking `TopLevelWidget`, but some for `SubWidget`.

**Reply by falkTX** [reply]
the only downside is that events are not handled the usual way.

so please clarify.. if you have a opengl window with contents already being drawn on it, and add a webview (as subwidget), how would that render on top of what is already there?
if we put a custom (sub)widget on top of the webview, does the drawing work correctly? (so the background of the window draws first, then the webview, then the custom (sub)widget).

If you can mix those things without issues, then webview can be a subwidget. otherwise it cannot.

**Reply by lucianoiam** [reply]
That's an interesting test, what do you think of the following?

1. Attach a TopLevelWidget to Window, or a SubWidget that can be configured to take up all the Window (I think parent class does not make a difference for this test?)
2. Add a WebWidget (the version that derives from SubWidget) as a child of #1 
3. Add a smaller (like a knob?) SubWidget as a child of #2
4. Check layout and input

If you can point me out some widgets ready for using in the above that would be a huge help, cannot find a centralized place and DPF-Widgets is very much WIP like it says.

I guess the experiment should be carried out for both OpenGL and Cairo?  btw, plugins do not compile when UI_TYPE=cairo on Windows, I guess that is expected.

**Reply by falkTX** [reply]
You can take any existing plugin from DPF, say Nekobi, and try to put a webview on the right side of the plugin.
If you construct the knobs after the webview, the knobs will be rendered after the webview.
Nekobi UI is quite small, but doesnt matter here much.

Nekobi, Mini-Series and ndc-Plugs can already be built and used with either cairo or opengl (opengl is used when available).
To force them to use cairo, build with `make HAVE_OPENGL=no`

**Reply by falkTX** [reply]
FYI there is a SubWidget::toFront() that you can use to bring some widget to the front of the rendering stack.

**Reply by lucianoiam** [reply]
Thanks, I hacked Nekobi and got some interesting results.

1. Unsurprisingly it is not possible to add subwidgets to the WebWidget. I made a silly assumption by thinking that every Widget was backed by a single native view, delegating composition to the platform. In turn there is a single OpenGL (or whatever) view that supports the entire DGL "canvas" so any foreign stuff added on top will obscure DGL, right?

<img width="1818" alt="Screen Shot 2021-06-21 at 19 08 00" src="https://user-images.githubusercontent.com/930494/122824790-6641e080-d2e1-11eb-9fe8-a1da4124a811.png">

So that will never work but also doesn't have any interesting use cases, all good.

2. Made the WebWidgets track `onResize()` and `onPositionChanged()` to update the native web view bounds as needed ( so it behaves like any other widget ) and the results are quite good:

- Linux: layout and input OK for both OpenGL and Cairo
- Mac: layout OK for both, input OK for OpenGL, input issues with Cairo (cannot tweak Nekobi knobs in the test)
- Windows: layout and input OK for OpenGL. Cannot compile with Cairo. I have some missing libs in MinGW and could never fix that.

Summarizing, I think this is not about `SubWidget` vs. `TopLevelWidget` anymore. Providing UI_TYPE=opengl, it is safe to say the `WebWidget` behaves similarly to `SubWidget` on all platforms except:

1. It cannot contain subwidgets
2. Its Z is always the highest and cannot be changed

Can you think of a way to enforce 1 and 2 at compile time?

Now that mixing widgets proved to work, can we build a web browser using DPF?  (joking)


https://user-images.githubusercontent.com/930494/122826866-f08b4400-d2e3-11eb-86c7-6cad8e2a100e.mov

https://user-images.githubusercontent.com/930494/122826875-f3863480-d2e3-11eb-90a9-ff3018bc93f6.mov

https://user-images.githubusercontent.com/930494/122826886-f719bb80-d2e3-11eb-8297-f9bdda1ac0e3.mov

**Reply by falkTX** [reply]
>     1. In turn there is a single OpenGL (or whatever) view that supports the entire DGL "canvas" so any foreign stuff added on top will obscure DGL, right?

That is correct. The webview being on top of the other widgets, even if rendered before or after, is not something we can change because this stack is from the OS-level.
The only way around it would be to create other OS-level embed windows/widgets/views and put those on top of the webview.
This is the kinda stuff I want to avoid adding into DPF, the complexities on this rise up pretty quickly..

I can understand wanting to position the web widget on a specific area of the window though.
Maybe we can have the webview still as top-level-widget, but have a special flag to tell the window to not resize it automatically for us. (though this behaviour should be optional, as it is useful to auto-resize to full width/height if for example UI type is html)

It just doesnt make sense to have this as a SubWidget if it is never really going to behave like one.

**Reply by lucianoiam** [reply]
Well actually for my own specific use case ( a plugin with a full blown HTML UI ), there is no need for web views < window area; this was more like an exercise and helped learning from DPF and refining the widgets. Repositioning is not probably very useful. However I think I am still missing something and that SubWidget vs TopLevelWidget is not only about dimensions. As far as I understand, many of the issues described also hold true if the widget was derived from TopLevelWidget, like not being able to contain subwidgets. What else would be needed to behave as a SubWidget?  just curious, I don't think I'll update the widgets much more now.

Agree on the complexity of embedding and IMO the use cases are a few even for general purpose frameworks.

**Reply by falkTX** [reply]
> As far as I understand, many of the issues described also hold true if the widget was derived from TopLevelWidget, like not being able to contain subwidgets

We can try to find workarounds for that, but yes that is still an issue.



> What else would be needed to behave as a SubWidget?

Integrating in the current running backend (opengl or cairo) most likely.
So if we need to do something like a scrolling page of widgets, to be able to.

In reality the webview is a foreign widget.


tbqh I would prefer to have the webview be only allowed if ui type is html, with the stub stuff, and not mix in any cairo or opengl.
but for that we still need that stub thing to work properly.

it all depends on what you want to do with it of course. if an opengl plugin wants to show some documentation built-in, I guess it makes sense.
but there is always caveats...

**Reply by lucianoiam** [reply]
> > As far as I understand, many of the issues described also hold true if the widget was derived from TopLevelWidget, like not being able to contain subwidgets
> 
> We can try to find workarounds for that, but yes that is still an issue.

IMO that could be just unsupported, it is very complex and unnecessary. The challenge is how to make it *elegantly* unsupported.

> 
> > What else would be needed to behave as a SubWidget?
> 
> Integrating in the current running backend (opengl or cairo) most likely.
> So if we need to do something like a scrolling page of widgets, to be able to.

True, the current SubWidget-based implementation would not behave properly in a scroll view.

> 
> In reality the webview is a foreign widget.
> 
> tbqh I would prefer to have the webview be only allowed if ui type is html, with the stub stuff, and not mix in any cairo or opengl.
> but for that we still need that stub thing to work properly.
> 
> it all depends on what you want to do with it of course. if an opengl plugin wants to show some documentation built-in, I guess it makes sense.
> but there is always caveats...

Yes undoubtedly this will never play completely well in the DGL arena. I would leave that up to you, how to eventually include the widget in DPF if finally decided to. For my own use case I'd say it already does more than I initially expected. Also there is a non-trivial aspect of the integration which is how to build the views. Each platform has some details needing attention, everything is in the dpf-webui Makefiles which are of acceptable quality but fitting it into the framework's own build scripts requires some effort.

**Reply by lucianoiam** [reply]
> Did you try to force gtk x11 backend?
> Since the gtk stuff is in a separate process, this wont conflict with any running tools.
> I know you can at least set an environment variable to set gtk backends, maybe possible with code too.

That's right. I did some research, it is also possible with code:

`gdk_set_allowed_backends("x11,wayland");`

The order defines which backend is tried first.

Now the GTK-based helper app thinks there's X11 but `XReparentWindow()` does not seem to work, the window to be embedded just disappears after the call. Nevermind, I don't have the time to deal with that. Thanks for the tip!

**Comment by falkTX** [comment]
So to be clear, it is best that webview is a top-level-widget, so we do not have to deal with child windows, partial embedding and stuff like that.
When using web stuff to render a widget, IMO it should be purely web/html and not mix cairo and opengl stuff with it.

Basically, think of Web/HTML as a target backend within DPF - So we get Cairo, OpenGL or Web/HTML.

**Reply by lucianoiam** [reply]
I think the difference between the two approaches is the scope of the web widget. Should it be a reusable component, or exclusively limited to a way of supporting completely web-based user interfaces? the dpf-webui project started with the latter goal in mind but then after some discussion, it took little effort to turn the web widget (90% of the project effort) into a reusable component.

**Comment by lucianoiam** [comment]
It's been some time since this, finally the project evolved into web UI support (JavaScript) and optionally web DSP as well (AssemblyScript). More or less at beta stage at this point [repo here](https://github.com/lucianoiam/hiphop). The result is pretty solid, there is a single thing I could still not fix but related to [WebKitGTK](https://bugs.webkit.org/show_bug.cgi?id=229227) and not DPF. There is a workaround in place but it breaks CSS viewport-based units on Linux.

The current webview version finally subclasses SubWidget for historic (read: lazy) reasons but now https://github.com/DISTRHO/DPF/issues/313 looks like the best approach.

Also just discovered https://github.com/LDPF , referencing relevant discussion here https://github.com/LDPF/DPF/commit/4cbed015de2d237b67a46621f631f9af932782ad

**Reply by falkTX** [reply]
Great stuff!
And yeah, the embed-external ui might be the best way around this. Before I thought that we might need the stub stuff, but on 2nd thought that is quite pointless since for example webgtk already creates its own X11 window.

**Comment by lucianoiam** [comment]
That's correct. There is a single caveat, maybe you know something about this. The GTK webview must be attached to a GtkWindow. Since the x11 window can already exist (ie, provided by the host) it is wrapped like this: GdkWindow* gdkChild = gdk_x11_window_foreign_new_for_display(gdk_display_get_default(), x11_window);

This works great but the webview does not seems to follow this special Gtk/GdkWindow resize events afterwards. Not to blame DPF, maybe you have experience wrapping x11 windows for use in Gtk and there are known limitations when "adopting" windows. In the meantime I filed a bug at the WebKitGTK tracker.

The current workaround is to create the webview with fixed large dimensions (like 2000x2000), it will be cropped by the wrapping window. And then use CSS for manipulating the body dimensions. As ugly as it sounds it works very good but it prevents vh/vw/etc from working correctly. This only applies to Linux.

**Reply by falkTX** [reply]
Dont you need a container to setup the sizes right?
https://github.com/lv2/suil/blob/master/src/x11_in_gtk2.c might have some useful code, not sure.
I dont typically deal with gtk.

**Comment by lucianoiam** [comment]
Thanks for the snippet will have a look at it. Resizing works as expected when creating a "normal" standalone GtkWindow and attaching the webview to it with no extra effort, it just works. There has to be a difference between a regular GtkWindow and the wrapped x11 version, already discovered that the latter does not emit some events maybe it is that. Not sure it is a supported setup though.

**Comment by lucianoiam** [comment]
Continuation of https://github.com/DISTRHO/DPF/pull/316

I like the idea of integrating all webview development into DPF. I started the project out of necessity as the first step of a project that needs a complex UI and runs on all desktop platforms. As a side effect, later discovered some overlap with the growing territory of web audio projects.

I see two main approaches:

- Only distribute the source files and hand over build process to the plugin dev. Like if they were reusable pseudo-widgets. Switching to external means that they are not real widgets anymore, in the DGL sense.
- Treat web as a new UI choice that depends on external. Have DPF take care of the build process like the current separate webview project does. This requires merging Makefiles and replicating things on CMake, I have no experience with the latter.

Right after both parts (external-ui and webview) are complete sounds like the right time to study how to merge. The current webview can already produce working plugins that depend on DGL [example here](https://github.com/lucianoiam/castello) , and there is a separate no-DGL branch that is already complete, minus standalone.

I would start by completing external-ui, then probably work a little more on the UI class as explained below to pave the way for integration.

Sharing some thoughts including long term ideas. The webview project provides two main classes, AbstractWebHostUI and AbstractWebView:

- AbstractWebHostUI: manages the web view, C++/JS bridging of some UI methods, and also provides  a minimal base for abstracting plugin vs standalone cases. Regarding the latter functionality, wouldn't it make sense to have that as a stock DPF feature? now we have external UI, I am quite sure other external UI devs will seek to abstract the plugin environment. For example, getParentWindowHandle() could return the host window or a minimal standalone window. I know this means repeating some DGL/Pugl code, but also potentially gives a more control to DPF. It would be something like UI_TYPE=native , or window. *Long term*:  in the future DGL widgets could run on top of this stack, instead of the current inverse situation.

- AbstractWebView: these were previously Widgets but does not make sense for external so the new nodgl branch renamed them to *WebView and stripped off any dependency on DGL. Ideally we would keep these generic, reusable even outside DPF.  *Long term*:  make them really generic, multiplatform webviews are very sought after and can benefit from a broader community contributions.  I think the only dependency on DPF right know is the String class

Then we have the WasmPluginHost but that is easier to handle because it does not require graphics. It is quite complete, including a subset of the DPF Plugin API ported to AssemblyScript. *Long term*: the only thing I am not sure about is keeping Wasmer as the Wasm engine, or switch to something minimal like WAMR. Time will tell, some stuff is even still discussed by the Wasm people.

Well there are some details to think about, tried to keep it short.
